lib/core/cpuMonitor.cpp

```diff
@@ -0,0 +1,114 @@
+#include "cpuMonitor.hpp"
+
+#include "Stage.hpp" // for Stage
+
+#include "json.hpp" // for basic_json<>::object_t, json, basic_json<>::value_type
+
+#include <chrono>      // for operator""ms, chrono_literals
+#include <functional>  // for _Bind_helper<>::type, _Placeholder, bind, _1, placeholders
+#include <math.h>      // for floor
+#include <pthread.h>   // for pthread_detach
+#include <stdio.h>     // for fclose, fopen, fscanf, snprintf, FILE
+#include <sys/types.h> // for pid_t
+#include <utility>     // for pair
+
+using namespace std::chrono_literals;
+using namespace std::placeholders;
+
+namespace kotekan {
+
+CpuMonitor::CpuMonitor() {
+    // Register CPU usage callback
+    restServer::instance().register_get_callback(
+        "/cpu_ult", std::bind(&CpuMonitor::cpu_ult_call_back, this, _1));
+}
+
+CpuMonitor::~CpuMonitor() {
+    restServer::instance().remove_get_callback("/cpu_ult");
+    ult_list.clear();
+}
+
+void CpuMonitor::start() {
+    this_thread = std::thread(&CpuMonitor::track_cpu, this);
+    pthread_detach(this_thread.native_handle());
+}
+
+void CpuMonitor::stop() {
+    stop_thread = true;
+}
+
+void CpuMonitor::track_cpu() {
+    while (!stop_thread) {
+        uint32_t cpu_times[10];
+        uint32_t cpu_time = 0;
+
+        // Read CPU stats from /proc/stat first line
+        std::string stat;
+        FILE* cpu_fp = fopen("/proc/stat", "r");
```

Comment:
Check return value for error conditions.

Reply:
This was marked as resolved, but you aren't checking the if `fopen` was successful.  This could cause a problem at the `fscanf` lines below if the file handle returned isn't valid.